ssl/statem/statem_srvr.c

```diff
@@ -1492,6 +1496,82 @@ MSG_PROCESS_RETURN tls_process_client_hello(SSL_CONNECTION *s, PACKET *pkt)
+     */
+    if (s == NULL)
+        return 0;
+    if (pkt == NULL) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    if (s->server == 1 && PACKET_remaining(pkt) != 0) {
+        int rv = 0, innerflag = -1;
+        size_t startofsessid = 0, startofexts = 0, echoffset = 0;
+        size_t outersnioffset = 0; /* offset to SNI in outer */
+        uint16_t echtype = OSSL_ECH_type_unknown; /* type of ECH seen */
+        const unsigned char *pbuf = NULL;
+
+        rv = ossl_ech_get_ch_offsets(s, pkt, &startofsessid, &startofexts,
+                                     &echoffset, &echtype, &innerflag,
+                                     &outersnioffset);
+        if (rv != 1) {
+            SSLfatal(s, SSL_AD_DECODE_ERROR, SSL_R_BAD_EXTENSION);
+            goto err;
+        }
+        if (innerflag == 1) {
+            WPACKET inner;
+
+            OSSL_TRACE(TLS, "Got inner ECH so setting backend\n");
+            /* For backend, include msg type & 3 octet length */
+            s->ext.ech.backend = 1;
+            s->ext.ech.attempted_type = TLSEXT_TYPE_ech;
+            OPENSSL_free(s->ext.ech.innerch);
+            s->ext.ech.innerch_len = PACKET_remaining(pkt);
+            if (PACKET_peek_bytes(pkt, &pbuf, s->ext.ech.innerch_len) != 1) {
+                SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+                goto err;
+            }
+            s->ext.ech.innerch_len += SSL3_HM_HEADER_LENGTH; /* 4 */
+            s->ext.ech.innerch = OPENSSL_malloc(s->ext.ech.innerch_len);
+            if (s->ext.ech.innerch == NULL)
+                goto err;
+            if (!WPACKET_init_static_len(&inner, s->ext.ech.innerch,
+                                         s->ext.ech.innerch_len, 0)
+                || !WPACKET_put_bytes_u8(&inner, SSL3_MT_CLIENT_HELLO)
+                || !WPACKET_put_bytes_u24(&inner, s->ext.ech.innerch_len
+                                          - SSL3_HM_HEADER_LENGTH)
+                || !WPACKET_memcpy(&inner, pbuf, s->ext.ech.innerch_len
+                                   - SSL3_HM_HEADER_LENGTH)
+                || !WPACKET_finish(&inner)) {
+                SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+                goto err;
+            }
+        } else if (s->ext.ech.es != NULL) {
+            PACKET newpkt;
+
+            if (ossl_ech_early_decrypt(s, pkt, &newpkt) != 1) {
+                SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+                goto err;
+            }
+            if (s->ext.ech.success == 1) {
+                /*
+                 * Replace the outer CH with the inner, as long as there's
+                 * space, which there better be! (a bug triggered a bigger
+                 * inner CH once;-)
+                 */
+                if (PACKET_remaining(&newpkt) > PACKET_remaining(pkt)
+                    || PACKET_replace(pkt, &newpkt) != 1) {
```

Comment:
Could you just do `PACKET_buf_init(pkt, PACKET_data(newpkt), PACKET_remaining(newpkt))` ?

Reply:
Yeah, that caused a bunch of test fails (in `tls_psk_do_binder()`) for early_data, which I guess might make sense, so not sure how to get around the ickkyness of `PACKET_replace`

As I said above, will ponder this more.

Reply:
I played about a bit more with that and still the same - the fact that the outer CH memory is inside a chunk allocated (I guess) for the entire record layer is a bit of a gotcha - using `PACKET_buf_init` seems liable to cause a leak at least ('cause I don't know how to free the buffer from here), and also seems to cause invalid reads in early data tests (I didn't quite bottom that out, but we end up with a wrong length and make valgrind unhappy;-).

So, I don't currently see anything better for now than that `PACKET_replace` thing.  (At least not without surgery on other layers of message processing, which I've not explored.)

Reply:
The why is because `pkt` doesn't point at the start of an allocated `PACKET` but is (IIRC) pointing inside one that contains the record layer message received. That includes the RL header and additional octets beyond the outer CH in the case of early data. (And that's also why `*pkt = *newpkt` won't work.)

Once I've addressed the other easy things, I can trace it out a bit more to verify the above - could be I guess that a smallish change higher up the call stack would mean one of these ideas would work. (That's the kind of change I've previously avoided but that might make sense now.)